### ISSUE TITLE
add_picture corruptes file in some case

### ISSUE THREAD

**Issue by Tores76** [python-openxml/python-docx#455]
Dear @scanny ,

First of all, thanks to you and all contributors for python-docx!

I have found a problem when I wanted to add images in an existing file. After the insertion the file resulted corrupted when opening it with MS Word.

I did some debugging I have found out that it was a clash between the id of the wp:docPr of the newly inserted image and the wp:docPr of a shape in the header.

I think that the problem is in next_id(self) of document.py that from my understanding looks only for the image ids in document.xml and not in header1.xml (where the conflicting wp:docPr is).
I have temporarily fixed it for my case by artificially increasing the id number of the newly inserted pictures by 10000 as I haven't find out as to access the header1.xml elements and I'm very short on time.

Again, thanks for this extremely useful package!

**Comment by scanny** [comment]
Thanks @Tores76. This is an interesting and useful finding. Your hypothesis makes sense to me; Word has access to the header whenever it has the document and I suppose it's reasonable they would place those ids into the same scope of uniqueness.

We don't have programmatic access to the header XML document(s) yet, so we wouldn't be able to search there yet. The artificial increase in ID for new ones we add makes sense to me though. I'll bet that's a pretty robust approach that would work in all cases we're likely to encounter.

I'll leave this open and put it on the short list for things we can fix next time we're in there.

**Comment by scanny** [comment]
@Tores76 By the way, can you post the code you used to increase the ID on new images you inserted? That could be very helpful to someone who needs it until we can get this into the build.

**Comment by Tores76** [comment]
Sure!
The solution I used at first was some kind of a one shot brute force in the code itself to check my hypothesis. Something like this:

from docx import Document
from docx.oxml.shared import qn

my_document = Document('without_images.docx')

# code inserting some images

doc_element = my_document._part._element
docPrs = doc_element.findall('.//' + qn('wp:docPr'))
for docPr in docPrs:
    docPr.set('id',str(int(docPr.get('id'))+100000))

my_document.save('with_images.docx')

And the output file was finally ok without having to remove the images in the header. A quick test with opc-diag seemed to show that when adding images with Word later on, it keeps using the lower unused ids without worrying about these high value ones.
This solution is straightforward and simple and may be all that is needed (it was for me) but is of course thought to be done once only at the end of all the insertions (the pool of ids could be considered messy if that step was to be repeated multiple times).

Now however I'm working with a modified property next_id in the document.py file:

    @property
    def next_id(self):
        """
        The next available positive integer id value in this document. Gaps
        in id sequence are filled. The id attribute value is unique in the
        document, without regard to the element type it appears on.
        """
        id_str_lst = self._element.xpath('//@id')
        used_ids = [int(id_str) for id_str in id_str_lst if id_str.isdigit() and int(id_str)>100000]
        for n in range(100001, len(used_ids)+100002):
            if n not in used_ids:
                return n
				
The idea is that all python-docx generated ids from this property live in have their own space at above 100000. It seems to work but unfortunately I haven't had the time yet to test it thoroughly in different scenario.

**Comment by scanny** [comment]
Fixed in release v0.8.7 circa Aug 18 2018.

**Comment by mrsanrise** [comment]
Get the same error in 0.8.10 release.
I use custom template with images in the footer and header.

word/header1.xml:wp:docPr id="13" 

word/document.xml:wp:docPr id="1" n
word/document.xml:wp:docPr id="2" n
word/document.xml:wp:docPr id="3" n
word/document.xml:wp:docPr id="4" n
word/document.xml:wp:docPr id="5" n
word/document.xml:wp:docPr id="6" n
word/document.xml:wp:docPr id="7" n
word/document.xml:wp:docPr id="8" n
word/document.xml:wp:docPr id="9" n
word/document.xml:wp:docPr id="10" 
word/document.xml:wp:docPr id="11" 
word/document.xml:wp:docPr id="12" 
word/document.xml:wp:docPr id="13" 

Docx is corrupted if I add greater than 12 images..

**Comment by LuisFilipeMLoureiro** [comment]
> Sure!
> The solution I used at first was some kind of a one shot brute force in the code itself to check my hypothesis. Something like this:
> 
> from docx import Document
> from docx.oxml.shared import qn
> 
> my_document = Document('without_images.docx')
> #code inserting some images
> doc_element = my_document._part._element
> docPrs = doc_element.findall('.//' + qn('wp:docPr'))
> for docPr in docPrs:
> docPr.set('id',str(int(docPr.get('id'))+100000))
> 
> my_document.save('with_images.docx')
> 
> And the output file was finally ok without having to remove the images in the header. A quick test with opc-diag seemed to show that when adding images with Word later on, it keeps using the lower unused ids without worrying about these high value ones.
> This solution is straightforward and simple and may be all that is needed (it was for me) but is of course thought to be done once only at the end of all the insertions (the pool of ids could be considered messy if that step was to be repeated multiple times).
> 
> Now however I'm working with a modified property next_id in the document.py file:
> 
> ```
> @property
> def next_id(self):
>     """
>     The next available positive integer id value in this document. Gaps
>     in id sequence are filled. The id attribute value is unique in the
>     document, without regard to the element type it appears on.
>     """
>     id_str_lst = self._element.xpath('//@id')
>     used_ids = [int(id_str) for id_str in id_str_lst if id_str.isdigit() and int(id_str)>100000]
>     for n in range(100001, len(used_ids)+100002):
>         if n not in used_ids:
>             return n
> ```
> 
> The idea is that all python-docx generated ids from this property live in have their own space at above 100000. It seems to work but unfortunately I haven't had the time yet to test it thoroughly in different scenarios.

Thanks for the code!  Can you elaborate how you call this function? Is the argument "self" the document.docx?

**Comment by LuisFilipeMLoureiro** [comment]
> Sure!
> The solution I used at first was some kind of a one shot brute force in the code itself to check my hypothesis. Something like this:
> 
> from docx import Document
> from docx.oxml.shared import qn
> 
> my_document = Document('without_images.docx')
> #code inserting some images
> doc_element = my_document._part._element
> docPrs = doc_element.findall('.//' + qn('wp:docPr'))
> for docPr in docPrs:
> docPr.set('id',str(int(docPr.get('id'))+100000))
> 
> my_document.save('with_images.docx')
> 
> And the output file was finally ok without having to remove the images in the header. A quick test with opc-diag seemed to show that when adding images with Word later on, it keeps using the lower unused ids without worrying about these high value ones.
> This solution is straightforward and simple and may be all that is needed (it was for me) but is of course thought to be done once only at the end of all the insertions (the pool of ids could be considered messy if that step was to be repeated multiple times).
> 
> Now however I'm working with a modified property next_id in the document.py file:
> 
> ```
> @property
> def next_id(self):
>     """
>     The next available positive integer id value in this document. Gaps
>     in id sequence are filled. The id attribute value is unique in the
>     document, without regard to the element type it appears on.
>     """
>     id_str_lst = self._element.xpath('//@id')
>     used_ids = [int(id_str) for id_str in id_str_lst if id_str.isdigit() and int(id_str)>100000]
>     for n in range(100001, len(used_ids)+100002):
>         if n not in used_ids:
>             return n
> ```
> 
> The idea is that all python-docx generated ids from this property live in have their own space at above 100000. It seems to work but unfortunately I haven't had the time yet to test it thoroughly in different scenarios.

Thanks! I am dealing with the same problem, but I am not being succesfull using the next_id fuction. Could you post the whole code of inserting imagens in an existing docx file and then using the fuction above?

**Comment by And2Devel** [comment]
> Thanks! I am dealing with the same problem, but I am not being succesfull using the next_id fuction. Could you post the whole code of inserting imagens in an existing docx file and then using the fuction above?

@LuisFilipeMLoureiro , simply replace function next_id in package (file docx/parts/story.py)

**Comment by janpieter-vanparys** [comment]
I think this still is a problem

**Comment by GauntPT** [comment]
> I think this still is a problem

This is correct. I have 0.8.11 and the following code fixed my images to be opened without any error message

tores76 fix helped: https://github.com/python-openxml/python-docx/issues/455#issuecomment-357748310

**Comment by jgrimard** [comment]
This issue should still be open in this repository.

**Comment by FgPy93** [comment]
@scanny the issue is still present in version 1.0.1 - Are there any problems with the proposed solution or why isn't it fixed yet? The proposal worked for me so far.

**Comment by scanny** [comment]
Reopening. Looks like `.next_id()` will need to check headers and footers as well as the document body for used image ids.

**Comment by maxmilianr** [comment]
Hi, i ran into the same issue in 1.1.2. Looking forward to the fix, and thanks for this package.

**Comment by jgrimard** [comment]
> Hi, i ran into the same issue in 1.1.2. Looking forward to the fix, and thanks for this package.

As a temporary solution to this issue I'm using python-docx-ng.  Specifically 0.9.3.  It broke in 0.9.4.  I haven't tested the later ones because 0.9.3 works well for my project.
https://pypi.org/project/python-docx-ng/0.9.3/

Hopefully it gets fixed in this repository so I can switch back.
